File: 3.cpp
// largest prime factor of 600851475143 . https://projecteuler.net/problem=3
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>
#include <numeric>
#include <utility>
#include <cmath>

typedef std::vector<int> iVec;
typedef std::vector<long long> llVec;

using std::cout;

bool is_prime(long long val) {
    long long ulimit = sqrt(val);
    for(long long i=2; i<ulimit; ++i) {
        if (val % i == 0)
            return false;
    }
    return true;
}

long long smallest_factor_and_larger_than(long long val, long long pre_start) {
    long long ulimit = sqrt(val);
    while(++pre_start <= ulimit) {
        if (val % pre_start == 0)
            return pre_start;
    }
    return val;
}
    

long long largest_prime_factor(long long val) {
    llVec factors;
    long long factor = 1;
    while(true) {
        factor = smallest_factor_and_larger_than(val, factor);
        factors.push_back(factor);
        if (factor == val)
            break;
    }
    long long res = 1;
    for(auto x: factors) {
        if(is_prime(x))
            res = x;
    }
    return res;
}

int main() {
    long long threshold = 600851475143;
    cout << largest_prime_factor(threshold);
}

File: 3.py
# largest prime factor of 600851475143 . https://projecteuler.net/problem=3
from math import sqrt

def is_prime(val):
    # val is prime iff (val%x != 0) for x in range(2,sqrt(val)).
    if val<2:
        return False
    for i in range(2, 1 + int(sqrt(val))):
        if val % i == 0:
            return False
    return True

def smallest_prime_factor(val):
    for i in range(2, int(sqrt(val))):
        if val % i == 0:
            return i
    return val

def smallest_factor_and_larger_than(val, pre_start=1):
    # return value could be non-prime!!
    for i in range(1+pre_start, int(sqrt(val))):
        if val % i == 0:
            return i
    return val

def largest_prime_factor(val):
    # condition of returned value: 1. prime 2. factor of val
    factors = []
    factor = 1
    while True:
        factor = smallest_factor_and_larger_than(val, factor)
        factors.append(factor)
        if factor == val:
            break
        val /= factor
    res = max(filter(is_prime, factors))
    return res

val = 600851475143
print(largest_prime_factor(val))
